File: src/api/companion/briefing.ts
/**
 * Morning Director briefing IPC — session-open composition + the
 * decision-ledger write every briefing action performs.
 *
 * Backend mirror: `src-tauri/src/commands/companion/briefing.rs` and
 * `src-tauri/src/companion/brain/briefing.rs` (serde camelCase — these
 * interfaces are hand-kept in sync; no ts-rs export on the Rust side).
 */
import { invokeWithTimeout as invoke } from '@/lib/tauriInvoke';

/** A persona that failed at least one run since the user left. */
export interface DeltaPersona {
  id: string;
  name: string;
  failedCount: number;
  /** False when already paused — pause actions against it are dropped. */
  enabled: boolean;
}

export interface DeltaAlert {
  ruleName: string;
  severity: string;
  message: string;
  personaId?: string | null;
}

export interface DeltaApproval {
  id: string;
  action: string;
  rationale: string;
}

/**
 * Serializable "what happened while you were away" document. Everything
 * the composer may reference — and everything an action may target — is
 * inside this doc; the backend validates composed actions against it.
 */
export interface SessionDelta {
  /** ISO timestamp of the previous session's end (last-seen anchor). */
  since: string;
  runs: number;
  failedRuns: number;
  alerts: number;
  approvalsWaiting: number;
  /** Open audit incidents (current state — the NOC feed). */
  openIncidents: number;
  failedPersonas: DeltaPersona[];
  alertSummaries: DeltaAlert[];
  pendingApprovals: DeltaApproval[];
}

/** Wire shape returned by `companion_compose_briefing`. */
export interface BriefingSpec {
  /** Serialized cockpit-spec body `{title, widgets, updated_at}`. */
  specJson: string;
  /** Provenance — `"athena"` for LLM-composed (sanitized) specs. */
  composedBy: string;
  generatedAt: string;
}

/**
 * Compose the morning briefing. Returns `null` when the backend gated
 * the call (trivial delta) or composition failed — callers render the
 * deterministic fallback instead. Generous timeout: this is a one-shot
 * LLM call.
 */
export async function companionComposeBriefing(
  delta: SessionDelta,
): Promise<BriefingSpec | null> {
  return invoke<BriefingSpec | null>(
    'companion_compose_briefing',
    { delta },
    { timeoutMs: 90_000 },
  );
}

/**
 * Write one briefing action to the decision audit ledger
 * (`companion_design_decision`, session `morning_briefing`) so the
 * decisions panel shows "what you did about it".
 */
export async function companionRecordBriefingAction(payload: {
  label: string;
  choice: string;
  rationale: string;
  personaContext?: string | null;
}): Promise<void> {
  return invoke<void>('companion_record_briefing_action', {
    label: payload.label,
    choice: payload.choice,
    rationale: payload.rationale,
    personaContext: payload.personaContext ?? null,
  });
}

File: src-tauri/src/commands/companion/briefing.rs
//! Tauri commands for the Morning Director briefing.
//!
//! `companion_compose_briefing` — the session-open composition entry.
//! The frontend computes the since-left [`SessionDelta`] and calls this
//! ONLY when the delta is non-trivial (the brain re-checks the gate).
//! Returns `None` when composition failed or was gated — the frontend
//! then renders its deterministic fallback briefing (the
//! `composeDefaultCockpit` model) instead of surfacing an error at the
//! worst possible moment (app open).
//!
//! `companion_record_briefing_action` — every action taken from a
//! briefing widget is written to the `companion_design_decision` audit
//! ledger, so the decisions panel shows "what you did about it" and the
//! trail is bidirectional (Athena proposes → user acts → both recorded).

use std::sync::Arc;

use serde::Serialize;
use tauri::State;

use crate::companion::brain::briefing::{self, SessionDelta};
use crate::companion::brain::decisions::{self, DecisionInput};
use crate::error::AppError;
use crate::ipc_auth;
use crate::AppState;

/// Session id stamped on briefing-action decision rows — groups them in
/// the audit trail without needing a chat session.
const BRIEFING_SESSION_ID: &str = "morning_briefing";

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BriefingSpec {
    /// Serialized cockpit-spec body (`{title, widgets, updated_at}`),
    /// same shape as `compose_cockpit` output; the frontend renders it
    /// as a contextual overlay.
    pub spec_json: String,
    /// Provenance: `"athena"` (LLM-composed, sanitized). The frontend's
    /// deterministic fallback never round-trips through here.
    pub composed_by: String,
    pub generated_at: String,
}

/// Compose the morning briefing from the frontend's session delta.
/// `Ok(None)` = gated or composition failed → deterministic fallback
/// client-side. Never surfaces the LLM error to the user.
#[tauri::command]
pub async fn companion_compose_briefing(
    state: State<'_, Arc<AppState>>,
    delta: SessionDelta,
) -> Result<Option<BriefingSpec>, AppError> {
    ipc_auth::require_auth(&state).await?;
    if briefing::delta_is_trivial(&delta) {
        return Ok(None);
    }
    match briefing::compose_briefing(&delta).await {
        Ok(spec_json) => Ok(Some(BriefingSpec {
            spec_json,
            composed_by: "athena".into(),
            generated_at: chrono::Utc::now().to_rfc3339(),
        })),
        Err(e) => {
            tracing::warn!(error = %e, "briefing compose failed; frontend falls back deterministically");
            Ok(None)
        }
    }
}

/// Record one briefing action into the decision audit ledger.
/// `label` = what was acted on ("Rerun failed persona"), `choice` = the
/// enum action kind taken, `rationale` = why (Athena's framing or the
/// deterministic fallback's), `persona_context` = target persona id /
/// approval id for `list_by_context` retrieval.
#[tauri::command]
pub async fn companion_record_briefing_action(
    state: State<'_, Arc<AppState>>,
    label: String,
    choice: String,
    rationale: String,
    persona_context: Option<String>,
) -> Result<(), AppError> {
    ipc_auth::require_auth(&state).await?;
    let input = DecisionInput {
        label: &label,
        choice: &choice,
        rationale: &rationale,
        decision_timestamp: None,
    };
    decisions::save_batch(
        &state.user_db,
        BRIEFING_SESSION_ID,
        persona_context.as_deref(),
        &[input],
    )?;
    Ok(())
}
